src/main/java/com/code_intelligence/jazzer/instrumentor/HookMethodVisitor.kt

```diff
@@ -92,6 +151,36 @@ private class HookMethodVisitor(
             return
         }
 
+        val skipHooksLabel = Label()
+        val applyHooksLabel = Label()
+        val useConditionalHooks = classWithHooksEnabledField != null
+        var postCallFrame: Pair<Array<Any>?, Array<Any>?>? = null
+        if (useConditionalHooks) {
+            val preCallFrame = (mv as? AnalyzerAdapter)?.let { storeFrame(it) }
+            // If hooks aren't enabled, skip the hook invocations.
+            mv.visitFieldInsn(
+                Opcodes.GETSTATIC,
+                classWithHooksEnabledField,
+                "hooksEnabled",
+                "Z",
+            )
+            mv.visitJumpInsn(Opcodes.IFNE, applyHooksLabel)
+            mv.visitMethodInsn(opcode, owner, methodName, methodDescriptor, isInterface)
+            postCallFrame = (mv as? AnalyzerAdapter)?.let { storeFrame(it) }
+            mv.visitJumpInsn(Opcodes.GOTO, skipHooksLabel)
+            if (preCallFrame != null) {
+                mv.visitFrame(
```

Comment:
This particular frame is not only required due to the instruction being the target of a jump, but also because it is the successor of an unconditional jump. Since the latter condition is not related to labels and the ASM `MethodVisitor` seems to visit frames before labels (https://asm.ow2.io/javadoc/org/objectweb/asm/MethodVisitor.html), I kept the order and added comments instead. What do you think?

Reply:
I might be mistaken here, but I've understood the [docs ](https://asm.ow2.io/javadoc/org/objectweb/asm/MethodVisitor.html#visitFrame(int,int,java.lang.Object%5B%5D,int,java.lang.Object%5B%5D)) differently. To my understanding, successors of unconditional branch instructions must only be visited if they are a target of a jump instruction or start an exception handler block. This would make sense since the successors of unconditional jumps are dead code if they are not the target of a jump instruction, and thus it wouldn't make sense to add a stack map frame unconditionally at these locations. I don't think that `visitLabel` adds any instructions, but only designates the instruction that will be visited just after it. That is why I think it is safe to move the `visitFrame` call to after the `visitLabel` call. That said, feel free to merge the code as is if my understanding is not correct.

Reply:
I always assumed that frames are also required on successors of unconditional branch instructions even if those aren't branch targets, possibly because the verifier doesn't want to track the reachability of every successor. But I don't know whether that is actually the case. 

While `visitLabel` doesn't emit any instructions, that doesn't automatically imply that the order of `visitLabel` and `visitFrame` calls doesn't matter for any of the (at least three) MethodVisitors involved. I looked into the implementation of ASM's `ClassReader` and it does seem to call `visitLabel` *after* visiting the instruction, which matches the javadocs. But when I do that here, I run into VerifyErrors. 

Given that `Frame, Insn, Label` doesn't seem to be working, `Label, Frame, Insn` is probably the next best order. I kept the comments, but reordered the statements to reflect that.